File: js_edition/src/absences.spec.js
import assert from 'assert';
import moment from 'moment';
import { absences, absencesAsIcal, getReason } from './api.js';

var inputs = new Map();
inputs.set('with no filter', absences());
inputs.set('with filter by user', absences(649));
inputs.set('with filter by date', absences(null, '2017-01-01', '2017-02-01'));

for (let [key, input] of inputs) {
  input.then((absences) => {
    describe('should export calendar with the same JSON data', () => {
      describe(key, () => {
        const result = absencesAsIcal(absences).toString();
        const matches = (result.match(/BEGIN:VEVENT/g) || []).length
        it('# of events', () => assert.equal(matches, absences.length));
        const event = absences[0].member.name + ' is ' + absences[0].reason;
        it('event name', () => assert(result.includes(event)));
        const startDate = moment(absences[0].startDate).utc().startOf('day').format('YYYYMMDD');
        it('event start date', () => assert(result.includes(startDate)));
        const endDate = moment(absences[0].endDate).utc().endOf('day').format('YYYYMMDD');
        it('event end date', () => assert(result.includes(endDate)));
      });
    })
  });
}

describe('should name absence reasons correctly', () => {
  it('sickness', () => assert.equal('sick', getReason('sickness')));
  it('vacation', () => assert.equal('on vacation', getReason('vacation')));
  it('something else', () => assert.equal('absent', getReason('something else')));
});

File: ruby_edition/spec/absences_spec.rb
require_relative '../cm_challenge/absences'

RSpec.describe 'Absences' do
  describe '#absences' do
    absences = CmChallenge::Absences.new

    inputs = [
      CmChallenge::Api.absences_with_members(),
      CmChallenge::Api.absences_with_members(649),
      CmChallenge::Api.absences_with_members(nil, '2017-01-01', '2017-02-01')
    ];

    inputs.each do |input|
      result = absences.to_ical(input).to_s
      # Test if we have the same amount of events in the calendar as we have in the input JSON
      it { expect(result.scan('BEGIN:VEVENT').length).to be(input.length) }

      input.each do |absence|
        event = "#{absence[:member][:name]} is #{absences.get_reason(absence[:type])}"
        start_date = absence[:start_date].tr('-', '')
        end_date = absence[:end_date].tr('-', '')

        # Test if the data in the calendar is the same as in the input JSON
        it { expect(result).to include(event) }
        it { expect(result).to include(start_date) }
        it { expect(result).to include(end_date) }
      end
    end

    # Test if we're naming the absence reasons properly
    it { expect(absences.get_reason('sickness')).to eq('sick') }
    it { expect(absences.get_reason('vacation')).to eq('on vacation') }
    it { expect(absences.get_reason('something else')).to eq('absent') }
  end
end
